instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelFutureInstrumentation.java

```diff
@@ -35,18 +38,74 @@
 
   @Override
   public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
-    return singletonMap(
+    Map<ElementMatcher.Junction<MethodDescription>, String> transformers = new HashMap<>();
+    transformers.put(
         isMethod()
             .and(named("addListener"))
             .and(takesArgument(0, named("io.netty.util.concurrent.GenericFutureListener"))),
         ChannelFutureInstrumentation.class.getName() + "$AddListenerAdvice");
+    transformers.put(
+        isMethod().and(named("addListeners")).and(takesArgument(0, isArray())),
+        ChannelFutureInstrumentation.class.getName() + "$AddListenersAdvice");
+    transformers.put(
+        isMethod()
+            .and(named("removeListener"))
+            .and(takesArgument(0, named("io.netty.util.concurrent.GenericFutureListener"))),
+        ChannelFutureInstrumentation.class.getName() + "$RemoveListenerAdvice");
+    transformers.put(
+        isMethod().and(named("removeListeners")).and(takesArgument(0, isArray())),
+        ChannelFutureInstrumentation.class.getName() + "$RemoveListenersAdvice");
+    return transformers;
   }
 
   public static class AddListenerAdvice {
     @Advice.OnMethodEnter
     public static void wrapListener(
-        @Advice.Argument(value = 0, readOnly = false) GenericFutureListener listener) {
-      listener = new WrappedFutureListener(Java8BytecodeBridge.currentContext(), listener);
+        @Advice.Argument(value = 0, readOnly = false)
+            GenericFutureListener<? extends Future<? super Void>> listener) {
+      listener = WrappedFutureListener.wrap(Java8BytecodeBridge.currentContext(), listener);
```

Comment:
I also looked at this briefly a few days ago. One thing that seemed problematic was that netty internally does instanceof on the listener type for example `progressiveListeners` in https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/DefaultPromise.java checks for `GenericProgressiveFutureListener`. Perhaps we should create a special wrapper for these to make instanceof pass. Hopefully there aren't more of these special listener types.

Reply:
Oh, that's an excellent catch 👍  Will implement.